fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/ai/tasks/VillagerTaskProvider.java

```diff
@@ -0,0 +1,616 @@
+	{
+		return applyToTasks(baseRandomTaskMap.get(taskType), profession, f);
+	}
+	
+	
+	public static List<Pair<Integer, ? extends Task<? super VillagerEntity>>> getBaseConstantTasks(TaskType taskType, /* @Nullable */ VillagerProfession profession, float f)
+	{
+		return applyToTasks(baseConstantTaskMap.get(taskType), profession, f);
+	}
+	
+	
+	
+	
+	
+	
+	public boolean hasRandomTasks(/* @NotNull */ TaskType taskType)
+	{
+		return randomTaskMap.containsKey(taskType);
+	}
+	
+	public boolean hasConstantTasks(/* @NotNull */ TaskType taskType)
+	{
+		return constantTaskMap.containsKey(taskType);
+	}
+	
+	public static boolean hasBaseConstantTasks(/* @NotNull */ TaskType taskType)
+	{
+		return baseConstantTaskMap.containsKey(taskType);
+	}
+	
+	public static boolean hasBaseRandomTasks(/* @NotNull */ TaskType taskType)
+	{
+		return baseRandomTaskMap.containsKey(taskType);
+	}
+	
+	public List<BiFunction<VillagerProfession, Float, Pair<Integer, ? extends Task<? super VillagerEntity>>>> getRawConstantTasks(TaskType taskType)
+	{
+		return constantTaskMap.get(taskType);
+	}
+	
+	public List<BiFunction<VillagerProfession, Float, Pair<Task<? super VillagerEntity>, Integer>>> getRawRandomTasks(TaskType taskType)
+	{
+		return randomTaskMap.get(taskType);
+	}
+	
+	public static List<BiFunction<VillagerProfession, Float, Pair<Integer, ? extends Task<? super VillagerEntity>>>> getRawBaseConstantTasks(TaskType taskType)
+	{
+		return baseConstantTaskMap.get(taskType);
+	}
+	
+	public static List<BiFunction<VillagerProfession, Float, Pair<Task<? super VillagerEntity>, Integer>>> getRawBaseRandomTasks(TaskType taskType)
+	{
+		return baseRandomTaskMap.get(taskType);
+	}
+	
+	
+	
+	
+	
+	
+	
+	
+	/*
+	 * Helper functions
+	 */
+	/* @NotNull */
+	private static <T> List<T> toListCheckNull(T[] arr)
+	{
+		if (arr == null)
+			return ImmutableList.of();
+		
+		return Arrays.asList(arr);
+	}
+	
+	private static <T> List<T> applyToTasks(List<BiFunction<VillagerProfession, Float, T>> l, final VillagerProfession p, final float f)
+	{
+		if (l == null)
+			return ImmutableList.of();
+		
+		return l.stream().map(el -> el.apply(p, f)).collect(Collectors.toList());
+	}
+	
+	private static <T> List<BiFunction<VillagerProfession, Float, T>> mapToBiFunction(final List<T> in)
+	{
+		final List<BiFunction<VillagerProfession, Float, T>> out = new ArrayList<>();
+		
+		for (final T t : in) {
+			if (t == null) {
+				continue;
+			}
+			
+			out.add((p, f) -> t);
+		}
+		
+		return out;
+	}
+	
+	private static <T> List<BiFunction<VillagerProfession, Float, T>> mapToBiFunction(final T[] in)
+	{
+		return mapToBiFunction(toListCheckNull(in));
+	}
+	
+	
+	/**
+	 * The type of {@code Task} being added.
+	 * @see TaskType#CORE
+	 * @see TaskType#WORK
+	 * @see TaskType#PLAY
+	 * @see TaskType#REST
+	 * @see TaskType#MEET
+	 * @see TaskType#IDLE
+	 * @see TaskType#PANIC
+	 * @see TaskType#PRERAID
+	 * @see TaskType#RAID
+	 * @see TaskType#HIDE
+	 */
+	public enum TaskType {
```

Comment:
This is problematic as it wouldn’t support adding an Activity to a villager brain. This would defenetly break a mod I would port over. Couldn’t we just track this stuff by using Activity instead of a enum TaskType?

Reply:
...I was not aware of that option.  Give me like 15 minutes.

Reply:
Huh, funny how my first implementation strategy - intercepting calls to `VillagerTaskListProvider` - ended up being the right option for implementing `Activity`.  What's even funnier is that I can't reuse my old code because too much has changed. :P
Well, at least it's just replacing Mixins this time and not scrapping 95% of my code to use an `EnumMap`, lol.